src/providers/route.js

```diff
@@ -0,0 +1,244 @@
+angular.module('bicker_router').provider('Route', function(ObjectHelper) {
+  "ngInject";
+  let tokens = {};
+  let urlWriters = {};
+  let urls = [];
+  let persistentStates = [];
+  let ready = false;
+  let types = {};
+  let html5Mode = false;
+
+  let provider = {
+
+    registerType(name, config) {
+      types[name] = config;
+      types[name].regex = new RegExp(types[name].regex.source, 'i');
+      return _.extend({ and: this.registerType }, this);
+    },
+
+    registerUrlToken(name, config) {
+      tokens[name] = _.extend({name}, config);
+      return _.extend({ and: this.registerUrlToken }, this);
+    },
+
+    registerUrlWriter(name, fn) {
+      urlWriters[name] = fn;
+      return _.extend({ and: this.registerUrlWriter }, this);
+    },
+
+    registerUrl(pattern, config) {
+      if (config == null) { config = {}; }
+      let urlData = {
+        compiledUrl: this._compileUrlPattern(pattern, config),
+        pattern
+      };
+
+      urls.push(_.extend(urlData, config));
+      return _.extend({ and: this.registerUrl }, this);
+    },
+
+    setPersistentStates(...stateList) {
```

Comment:
More self-executing anonymous functions... I'm not going to continue to comment, please find and replace them all